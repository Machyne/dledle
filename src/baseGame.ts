// Give these string values so they can be round trip through being used as object keys.
// These are never encoded as-is so strings are fine.
export enum GameScore {
  Win = "Win",
  NearWin = "NearWin",
  Loss = "Loss",
}

export type GameResult = {
  score: GameScore;
  serializedResult: string;
};

export abstract class BaseGame {
  private _resultRegex: RegExp | null = null;

  /**
   * The name of the game.
   */
  abstract get name(): string;

  /**
   * The name of the game in lowercase with only alphanumeric characters.
   */
  get fileName(): string {
    return this.name.toLowerCase().replace(/[^a-z0-9]/g, "");
  }

  /**
   * A regex that matches the result of the game.
   */
  get resultRegex(): RegExp {
    if (this._resultRegex === null) {
      this._resultRegex = this._buildResultRegex();
    }
    return this._resultRegex;
  }

  /**
   * Builds a regex that matches the result of the game. The return value will be cached.
   */
  abstract _buildResultRegex(): RegExp;

  /**
   * A link to the game.
   */
  abstract get link(): string;

  /**
   * Parses the user-supplied game result and returns a serialized version of
   * the game data in addition to the "score".
   *
   * @param gameResult the result of the game run through the resultRegex
   */
  abstract serializeResult(gameResult: RegExpMatchArray): GameResult;

  /**
   * The share text of the result of the game given the serialized game data.
   *
   * @param serializedResult the serialized game data
   */
  abstract deserialize(serializedResult: string): string;
}
