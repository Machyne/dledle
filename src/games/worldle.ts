import { BaseGame, GameScore } from "../baseGame";
import { DEFAULT_WORD_WIDTH, intToBase64, nextInt } from "../util/base64";
import {
  emojiToRegexUnion,
  fiveSquaresRegex,
  ordinalArrowValues,
  ordinalArrows,
  splitEmojiLines,
} from "../util/emoji";
import { emojiSquareScaleToNumber, numberToEmojiSquareScale } from "../util/gameHelpers";

const winEmoji = "ðŸŽ‰";
const arrowsOrWin = [...ordinalArrows, winEmoji];
const arrowOrWinValues: Record<string, number> = {
  ...ordinalArrowValues,
  "ðŸŽ‰": ordinalArrows.length,
};

export class Worldle extends BaseGame {
  get name(): string {
    return "Worldle";
  }

  get link(): string {
    return "https://worldle.teuteuf.fr/";
  }

  _buildResultRegex(): RegExp {
    const guessStr = fiveSquaresRegex + emojiToRegexUnion(arrowsOrWin);
    return new RegExp(
      `#Worldle #(\\d+) ([1-6X])\\/6 \\((\\d{1,3}%)\\)\\s+((?:${guessStr}\\s*){1,6})`,
      "u",
    );
  }

  serializeResult(gameResult: RegExpMatchArray) {
    const [, gameNumber, numGuessesStr, percentStr, guessesBlock] = gameResult;
    const isWin = numGuessesStr !== "X";

    // First encode the game number, num guesses, and percent score.
    const percent = parseInt(percentStr);
    let encoded = intToBase64(parseInt(gameNumber));
    const numGuesses = isWin ? parseInt(numGuessesStr) : 6;
    encoded += intToBase64(percent * 12 + (numGuesses - 1) * 2 + (isWin ? 1 : 0));
    const guesses = splitEmojiLines(guessesBlock);
    if (guesses.length !== numGuesses) {
      throw new Error(`Expected ${numGuesses} guesses, got ${guesses.length}`);
    }
    // Each guess is a closeness value (0-10) and a direction (0-8). We can fit two of these into
    // 3 base64 chars, since (11*9)^2 < 64^3.
    let guessEncoding = 0;
    for (let guessIdx = 0; guessIdx < guesses.length; ++guessIdx) {
      const guessChars = guesses[guessIdx];
      const closeness = emojiSquareScaleToNumber(guessChars);
      const direction = arrowOrWinValues[guessChars[5]];
      let rowEncoding = closeness * 9 + direction;
      if (guessIdx % 2 === 1) {
        rowEncoding *= 11 * 9;
      }
      guessEncoding += rowEncoding;

      // Write the encoding every other row, or at the end.
      if (guessIdx % 2 === 1 || guessIdx === guesses.length - 1) {
        encoded += intToBase64(guessEncoding);
        guessEncoding = 0;
      }
    }
    const score = isWin ? GameScore.Win : percent >= 90 ? GameScore.NearWin : GameScore.Loss;
    return { score, serializedResult: encoded };
  }

  deserialize(serializedResult: string) {
    if (
      serializedResult.length % DEFAULT_WORD_WIDTH !== 0 ||
      serializedResult.length < DEFAULT_WORD_WIDTH * 3
    ) {
      throw new Error(`Invalid Worldle result: ${serializedResult}`);
    }
    // Start with the metadata.
    let intResult: number;
    [intResult, serializedResult] = nextInt(serializedResult);
    const gameNumber = intResult;
    [intResult, serializedResult] = nextInt(serializedResult);
    const isWin = intResult % 2;
    intResult = Math.floor(intResult / 2);
    const numGuesses = (intResult % 6) + 1;
    intResult = Math.floor(intResult / 6);
    const percent = intResult;

    // Then decode the guess rows.
    const guesses: string[] = [];
    while (serializedResult.length > 0) {
      [intResult, serializedResult] = nextInt(serializedResult);
      const firstGuess = intResult % (11 * 9);
      const secondGuess = Math.floor(intResult / (11 * 9));
      const makeGuess = (guess: number) => {
        const direction = guess % 9;
        const closeness = Math.floor(guess / 9);
        return numberToEmojiSquareScale(closeness) + arrowsOrWin[direction];
      };
      guesses.push(makeGuess(firstGuess));
      const hasSecondGuess = numGuesses > guesses.length;
      if (hasSecondGuess) {
        guesses.push(makeGuess(secondGuess));
      }
    }
    return `#Worldle #${gameNumber} ${isWin ? numGuesses : "X"}/6 (${percent}%)\n${guesses.join(
      "\n",
    )}`;
  }
}
