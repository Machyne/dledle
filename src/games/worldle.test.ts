import { GameScore } from "../baseGame";
import { Worldle } from "./worldle";

const sampleOddRowWin = `
#Worldle #700 5/6 (100%)
🟩🟩🟩🟨⬛↗️
🟩🟩🟩🟩⬛↗️
🟩🟩🟩🟩⬛⬆️
🟩🟩🟩🟩🟨↗️
🟩🟩🟩🟩🟩🎉
`;

const sampleEvenRowWin = `
#Worldle #701 2/6 (100%)
🟩🟩🟩🟩⬛⬇️
🟩🟩🟩🟩🟩🎉
`;

const sampleNearWin = `
#Worldle #701 X/6 (92%)
🟩🟩🟩🟩⬛↙️
🟩🟩🟩🟨⬛⬆️
🟩🟩🟩🟨⬛↘️
🟩🟩🟩🟩🟨↘️
🟩🟩🟩⬛⬛➡️
🟩🟩🟩🟨⬛⬅️
`;

const sampleLoss = `
#Worldle #123 X/6 (81%)
🟩⬛⬛⬛⬛↙️
🟩🟨⬛⬛⬛⬆️
🟩🟩🟩⬛⬛↘️
🟩🟨⬛⬛⬛↘️
🟩🟩🟩⬛⬛➡️
🟩🟩🟩⬛⬛⬅️
`;

describe("Worldle", () => {
  const wordle = new Worldle();

  it("can parse with extra link text", () => {
    const match = wordle.resultRegex.exec(sampleOddRowWin + "\nhttps://worldle.teuteuf.fr");
    expect(match).not.toBeNull();
    const { score, serializedResult } = wordle.serializeResult(match!);
    expect(score).toEqual(GameScore.Win);
    expect(wordle.deserialize(serializedResult)).toEqual(sampleOddRowWin.trim());
  });

  for (const [name, input, expectedScore] of [
    ["sampleEvenRowWin", sampleEvenRowWin, GameScore.Win],
    ["sampleOddRowWin", sampleOddRowWin, GameScore.Win],
    ["sampleNearWin", sampleNearWin, GameScore.NearWin],
    ["sampleLoss", sampleLoss, GameScore.Loss],
  ] as Array<[string, string, GameScore]>) {
    it(`should serialize and deserialize ${name} correctly`, () => {
      const match = wordle.resultRegex.exec(input);
      expect(match).not.toBeNull();
      const { score, serializedResult } = wordle.serializeResult(match!);

      expect(score).toEqual(expectedScore);
      expect(wordle.deserialize(serializedResult)).toEqual(input.trim());
    });
  }
});
