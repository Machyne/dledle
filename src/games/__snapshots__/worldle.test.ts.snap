// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Worldle serialized data should match snapshot 1`] = `
{
  "sampleEvenRowWin": {
    "score": "Win",
    "serializedResult": "9KAzSAwYC",
  },
  "sampleEvenRowWin trimmed": {
    "score": "Win",
    "serializedResult": "9KAzSAwYC",
  },
  "sampleEvenRowWin with extra link text": {
    "score": "Win",
    "serializedResult": "9KAzSAwYC",
  },
  "sampleLoss": {
    "score": "Loss",
    "serializedResult": "7BAWPAJqAbyACZB",
  },
  "sampleNearWin": {
    "score": "NearWin",
    "serializedResult": "9KAaRAriBGGC9mB",
  },
  "sampleOddRowWin": {
    "score": "Win",
    "serializedResult": "8KA5SAn2BnECiBA",
  },
}
`;
